### DISCUSSION TITLE
CoreWCF v1.4.0-preview1

### DISCUSSION THREAD

**Discussion by mconnew** [CoreWCF/CoreWCF#1069] [Releases]
# Release Notes

## Package Info

[CoreWCF.Primitives 1.4.0-preview1](https://www.nuget.org/packages/CoreWCF.Primitives/1.4.0-preview1)  
[CoreWCF.Http 1.4.0-preview1](https://www.nuget.org/packages/CoreWCF.Http/1.4.0-preview1)  
[CoreWCF.WebHttp 1.4.0-preview1](https://www.nuget.org/packages/CoreWCF.WebHttp/1.4.0-preview1)  
[CoreWCF.NetFramingBase 1.4.0-preview1](https://www.nuget.org/packages/CoreWCF.NetFramingBase/1.4.0-preview1)
[CoreWCF.NetTcp 1.4.0-preview1](https://www.nuget.org/packages/CoreWCF.NetTcp/1.4.0-preview1)  
[CoreWCF.NetNamedPipe 1.4.0-preview1](https://www.nuget.org/packages/CoreWCF.NetNamedPipe/1.4.0-preview1)
[CoreWCF.Queue 1.4.0-preview1](https://www.nuget.org/packages/CoreWCF.Queue/1.4.0-preview1)
[CoreWCF.MSMQ 1.4.0-preview1](https://www.nuget.org/packages/CoreWCF.MSMQ/1.4.0-preview1)
[CoreWCF.RabbitMQ 1.4.0-preview1](https://www.nuget.org/packages/CoreWCF.RabbitMQ/1.4.0-preview1)
[CoreWCF.RabbitMQ.Client 1.4.0-preview1](https://www.nuget.org/packages/CoreWCF.RabbitMQ.Client/1.4.0-preview1)
[CoreWCF.ConfigurationManager 1.4.0-preview1](https://www.nuget.org/packages/CoreWCF.ConfigurationManager/1.4.0-preview1)
[CoreWCF.Templates 1.4.0-preview1](https://www.nuget.org/packages/CoreWCF.Templates/1.4.0-preview1)

## .NET Compatibility
This release depends on .NET Standard 2.0 and runs on any .NET version which supports .NET Standard 2.0. This release supports .NET Framework 4.6.2 and above, .NET 6, and .NET 7. It is built on top of ASP.NET Core and has been tested and runs on all currently supported versions of ASP.NET Core up to 7.0. The CoreWCF.RabbitMQ.Client package only supports .NET 6 or later.

### Named Pipe support
We've added the CoreWCF.NetNamedPipe package to provide support for using named pipes. The code common to NetNamedPipe and NetTcp has been refactored into the CoreWCF.NetFramingBase. This makes it easier to add new transports based on top of streaming connections. Named Pipe support is only for Windows, but we'll be adding Unix Domain Socket support at a later date to provide an equivalent feature for using on Linux. This was contributed by @mconnew from Microsoft.   

### Queue Transport support
We've added the CoreWCF.Queue package which provides a base set of classes to make adding Queue based transports easier. We are releasing the new packages CoreWCF.MSMQ and CoreWCF.RabbitMQ (implements the AMQP protocol) to provide transports for using Windows MSMQ and RabbitMQ. We also currently have in development a queue based transport for Apache Kafka which will be released in the near future. The goal is to provide support for MSMQ as a step to migrating to other queue technologies. We're also releasing our first client package CoreWCF.RabbitMQ.Client which is only supported on .NET 6.0 or later. This is built on top of the WCF Client 6.0 pre-release package. This work was coordinated by @birojnayak from Amazon AWS, who wrote the design [document](https://github.com/CoreWCF/CoreWCF/blob/main/Documentation/DesignDocs/corewcfqueue-design.md) for the generic Queue implementation. The majority of the code was contributed by @Ximik87 and @jonlouie. We also had code contributions from @birojnayak and @g7ed6e.  

### Add support for ServiceBehaviorAttribute.UseSynchronizationContext
@bjornhellander exposed `ServiceBehaviorAttribute.UseSynchronizationContext` and added validating tests to confirm the property is working as expected. #983

### Bug Fixes
When using `HttpClientCredentialType.InheritedFromHost`, we now issue an auth challenge if request is not authenticated if earlier middleware didn't do so, by @g7ed6e in #967  
Fixed DI injection generated code when parameter name in implementation doesn't match parameter name in contract, by @g7ed6e in #1030  
Fixed channel state transitions for HTTP which was breaking returning a fault when user extensibility throws an exception, by @mconnew in #1011  
Fixed MTOM headers being incorrectly added to response body with HTTP, by @mconnew in #1049  
Fixed compat behavior of `[WebGet|WebInvoke]Attribute.Is[Request|Response]FormatSetExplicitly` when using the attributes from `System.ServiceModel.dll` when using WebHttp on .NET Framework, by @arontsang in #1035  

### Other changes
The version of dependent packages has changed in this release. We've had feedback that depending on the latest versions of packages causes problems for some users as it forces an upgrade. Starting from this release, we'll depend on the latest LTS versions of packages. This means we've downgraded requirements for a couple of dependencies from the 7.x version to the 6.x version. Upgrading and using the 7.x versions of packages in your own projects is still supported and intended to work.  

As part of the work necessary to refactor common code between NetTcp and NetNamedPipe, CoreWCF now supports hosting NetTcp services when using an HTTP IServer other than Kestrel. This means you can use NetTcp with HttpSys or even hosted in IIS. Please note, NetTcp still doesn't support IIS hosted activation or port sharing, it still listens on its own socket and needs a unique port number.

### Non-shipping contributions
None

### New Contributors
* @bjornhellander made their first contribution in https://github.com/CoreWCF/CoreWCF/pull/983

### Feedback
Your feedback is important and appreciated. Please use the discussion https://github.com/CoreWCF/CoreWCF/discussions/1069 for your questions and comments.

**Full Changelog**: https://github.com/CoreWCF/CoreWCF/compare/v1.3.2...v1.4.0-preview1

<hr /><em>This discussion was created from the release <a href='https://github.com/CoreWCF/CoreWCF/releases/tag/v1.4.0-preview1'>CoreWCF v1.4.0-preview1</a>.</em>

**Comment by eaexo** [comment]
Hello @mconnew, @tibor19 

What are the current plans/ETA to make CoreWCF 1.4.0 a release (and hence production ready)?
What is the current of CoreWCF 1.4.0 preview? Is it safe to use it in production, or should we wait for official release?

Thanks Evgeni

**Reply by mconnew** [reply]
I'm heads down on ReliableSessions because of an internal high priority ask so haven't had time to come back to the preview release.  
It was in preview for 3 reasons.
1. New base Queue api design needed some time to validate - I'm happy the design is correct now.
2. Lots of new code for the NetNamedPipe implementation - 3 issues have been found, only one of them is a production blocker and affects NetNamedPipe with streamed mode transfer. The other two, one has a workaround, the other is a failure scenario fails the wrong way. We don't have WSDL support for it yet, and the service shutdown isn't very graceful. As long as you aren't using streamed mode transfer, I haven't heard anything to suggest it's not stable. 95% of the code is common to the existing NetTcp codebase which helps with stability.
3. MSMQ needs to be reworked to not depend on 3rd party non-supported package.

We'll do a partial release keeping some pieces still preview soon, but I'm spending all my time on the ReliableSessions implementation at the moment. The main difference from preview to release is in Microsoft official support. If you don't have a paid support contract, then there isn't much difference for you. Other than the issues that were found and the lack of WSDL/clean shutdown, the code is ready for release.
